### PR TITLE
Add SSL related params to ClientSession.request

### PR DESCRIPTION
**Pull request by cecton**
## What do these changes do?

...

## Are there changes in behavior for the user?

...

## Related issue number

#1128 

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `changes` folder
  * name it `<issue_id>.<type>` for example (588.bug)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by cecton** [comment]
Having issues with too much mocking.... if someone has any advice?

**Comment by asvetlov** [comment]
I suggest the following:

1. Make a self signed certificate and push it into `tests` (or just reuse `tests/sample.crt` if it's good for you).
2. Write functional tests using real app started with self signed certs.

Feel free to ask if my recommendation is not clean.

P.S.
Very thank you for picking up the issue and working on implementation.

**Comment by asvetlov** [comment]
P.P.S.
Side note: as committer you are free to drop github template for new issues and PRs.
These templates are created for helping newbies to file PR properly.
You know our rules already and might use free form for issue/PR description.
For example PR checklist could be skipped if you are keeping the list in your memory.
The only requirement: if PR is related to existing issue please mention it in description text -- it simplifies navigation between issue and PR.

**Comment by asvetlov** [comment]
Obviously it's up to you: use free form or file all template fields if template is more comfortable to you.
Personally I prefer free form for both.

**Comment by asvetlov** [comment]
Regarding to proxy tests: take a look on `tests/proxy_functional.py` design.

**Comment by cecton** [comment]
@asvetlov PR template is very fine for me :) it helped me a lot to not forget anything (documentation and change logs especially) but thx

I will have to change a few (all?) tests that mock the entire ClientRequest class. This thing is settings properties in the ClientRequest that I don't want.

**Comment by asvetlov** [comment]
Are you talking about proxy tests which should be affected by the PR, not fingerprint related tests, right?

In this case please don't hesitate to update all mocks if method signatures changed. I see no way to do it better.

**Comment by cecton** [comment]
Yes at this point the code I did should have kept the original behavior intact and the tests should pass unchanged. But 3 of the 4 failures are due to the fact that the mocking is actually giving a fingerprint to the ClientRequest objects.

**Comment by asvetlov** [comment]
Please update failed mocks, it's totally correct.

**Comment by asvetlov** [comment]
@cecton would you continue work on the PR?
I have a feeling that we need a new aiohttp release, will dedicate a full time week to lift up pending issues.

**Comment by cecton** [comment]
@asvetlov it shouldn't have take that much time but I got stuck with the mocking thing and then I had a lot of changes recently (change of project basically).

I was going to ask for help with the tests actually.

**Comment by cecton** [comment]
@asvetlov okay ready for review

**Comment by asvetlov** [comment]
Looks good. Feel free to merge, while I very appreciate full test coverage.

**Comment by cecton** [comment]
@asvetlov Thanks I will improve that! I thought I had all the cases but I hadn't check the diff on codecov.

**Comment by asvetlov** [comment]
Just a hint: install codecov browser extension: https://docs.codecov.io/v4.3.6/docs/browser-extension
It allows to see coverage just in github 'Files Changes' tab.
If you'll update contributors instruction to mention the plugin -- it would be awesome!

**Comment by jettify** [comment]
that is nice trick!

File: aiohttp/client_reqrep.py [hunk 1]
```diff
@@ -307,6 +328,35 @@ def update_proxy(self, proxy, proxy_auth, proxy_from_env):
...
         self.proxy_auth = proxy_auth
 
+    def update_fingerprint(self, fingerprint):
+        if fingerprint:
+            digestlen = len(fingerprint)
+            hashfunc = HASHFUNC_BY_DIGESTLEN.get(digestlen)
+            if not hashfunc:
+                raise ValueError('fingerprint has invalid length')
+            elif hashfunc is md5 or hashfunc is sha1:
+                warnings.simplefilter('always')
```
Comment (kxepal) [inline aiohttp/client_reqrep.py]: Wouldn't this cause suddenly global side effect?

Comment (cecton) [reply aiohttp/client_reqrep.py]: I have no idea. I copy-pasted that from there a610cf12209326fba5e6660c1644cbff55b7c07d

Comment (cecton) [reply aiohttp/client_reqrep.py]: I think you're right. After this the "warning filter specifications" is altered. https://docs.python.org/3.5/library/warnings.html#warnings.simplefilter @asvetlov can you confirm you want to keep this behavior?

**Comment by asvetlov** [comment]
@jettify 
> that is nice trick!

What are you talking about?

File: aiohttp/client_reqrep.py (hunk 1 above)
Comment (asvetlov) [reply aiohttp/client_reqrep.py]: No.
@kxepal good catch.
Please don't alter `warnings` filter but duplicate the message to `aiohttp.log.client_logger`.
From my experience people are much more sensible to log messages than deprecation warnings.
Log level should be [`warning'](https://docs.python.org/3/library/logging.html#logging.Logger.warning) sure.

**Comment by jettify** [comment]
@asvetlov codecov browser extension, very usefull

**Comment by asvetlov** [comment]
@jettify thanks

File: aiohttp/client_reqrep.py (hunk 1 above)
Comment (cecton) [reply aiohttp/client_reqrep.py]: I suppose I change the one in connector too? ^_^

**Comment by cecton** [comment]
@asvetlov I just saw that the original ticket mention a "move" of those parameters. She I implement a deprecation for the using the parameter in the connector?

**Comment by asvetlov** [comment]
I not sure.
Obviously there is a need for per-request ssl configuration but I don't know is deprecation makes sense.
Maybe yes.
Guys what do you think?

**Comment by cecton** [comment]
@asvetlov the fingerprint is something you will get per server basis so per request makes sense. On the other hand the connector has the argument fingerprint like it would connect always to the same server. To me it doesn't make any sense. Unless maybe for a UNIX socket?

`verify_ssl` and `ssl_context` could make sense because you may want to disable the entire SSL checking for a complete ClientSession.

**Comment by asvetlov** [comment]
Make sense

**Comment by asvetlov** [comment]
Thank you!

**Pull request merged**